### ISSUE TITLE
Hatching in EPS figures doesn't show up in Preview on MacOS X with Matplotlib 2.0

### ISSUE THREAD

**Issue by astrofrog** [matplotlib/matplotlib#8289]
Patches that have a hatch pattern in EPS plots don't show the hatch pattern when opened in Preview on MacOS X. This may be a Preview bug rather than a Matplotlib bug, but I wanted to check whether anything can be done on the Matplotlib side to avoid this (since the likelihood Apple will care about this issue is minimal). This was not a problem with Matplotlib 1.5, only with 2.0, so in that sense this is a regression.

To reproduce this:

```python
from matplotlib import pyplot as plt
from matplotlib.patches import Ellipse

fig = plt.figure()
ax = fig.add_subplot(1,1,1)
ell = Ellipse((0.5, 0.5), 0.3, 0.4, 30, hatch='/', facecolor='none', edgecolor='black')
ax.add_patch(ell)
fig.savefig('test.eps')
```

**PNG:**

![test](https://cloud.githubusercontent.com/assets/314716/23892640/595fe6fe-0893-11e7-82a5-ac8f49fd2e48.png)

**EPS, as viewed in Preview on MacOS X:**

<img width="885" alt="screen shot 2017-03-14 at 08 52 50" src="https://cloud.githubusercontent.com/assets/314716/23892976/aab18908-0894-11e7-817b-a8a8a811a424.png">

The above is with Matplotlib 2.0. With 1.5, Preview correctly shows the hatch pattern:

<img width="550" alt="screen shot 2017-03-14 at 09 01 54" src="https://cloud.githubusercontent.com/assets/314716/23893052/ef116d48-0894-11e7-80ab-001f3d306f2e.png">

I'm using MacOS X 10.11.6

**Comment by afvincent** [comment]
Just to be sure: the hatches do actually exist in the EPS file (e.g. the issue is only with the preview)? I ran your MWE on Linux and Python 2 with Matplotlib 1.5.1 and 2.0: in Evince I can see hatches in both of them. However the hatches do not look exactly identical between both records, which I guess is due to #7421 (ping @QuLogic)

**Comment by astrofrog** [comment]
@afvincent - yes, I think they do, I can see the hatches if I open the file with ghostscript

**Comment by tacaswell** [comment]
what is the diff between the two files?

**Comment by QuLogic** [comment]
The changes in the individual commits of #7421 are fairly self-contained; it should be possible to try reverting them one at a time to see if they have any effect. At a guess, I'd say [this commit](https://github.com/matplotlib/matplotlib/pull/7421/commits/f2853bd696c39c63fb2020ce4766173a5d7f6b9b) and Preview has some trouble with the `setpattern` shortcut.

**Comment by tacaswell** [comment]
https://github.com/matplotlib/matplotlib/pull/7421/files#diff-b465967f537465fec4960b152fa49cffR903 is likely the specific change at fault.

https://www.adobe.com/products/postscript/pdfs/PLRM.pdf has the documentation on this and 

>setpattern is the normal method  for  selecting  patterns;  in  practice,  it  is  rarely  necessary  to  set  the  color space and color value separately

pg 264

**Comment by afvincent** [comment]
Sorry, unexpected workload at day-job last night. So here are the differences that I can observe both in Evince and GS between Matplotlib 1.5.1 and 2.0:
![eps_and_hatches_differences_between_1 5 1_and_2 0](https://cloud.githubusercontent.com/assets/17270724/23938998/07c9914c-095f-11e7-9cd8-441e1bf4e2ba.png)
The main one is that the hatches do not have a constant line width with 2.0.

**Comment by WeatherGod** [comment]
Well, one difference is that the default figure size changed between 1.5
and 2.0. The spacing of the hatches is defined in points, so changing the
figure size will change that. I would suspect that some other changes to
the default style might explain some of the other 1.5/2.0 differences.

On Wed, Mar 15, 2017 at 4:11 AM, Adrien F. Vincent <notifications@github.com

**Comment by afvincent** [comment]
@WeatherGod I think the problem is not about the hatching style but more about how hatching is defined when saving in postscript and how it affects the rendering in Preview, Evince, GS, etc. as @QuLogic and @tacaswell have suggested. For example, when the hatches density is doubled ('//' instead of '/'), the not very aesthetic bold segments on some of the hatches are still there in EPS produced by 2.0, while the line width stays smooth and constant with 1.5(.1):
![eps_and_hatches_differences_between_1 5 1_and_2 0_bis](https://cloud.githubusercontent.com/assets/17270724/23989482/eb043dee-0a32-11e7-9735-a11bdd6e4a2c.png)

**Comment by tacaswell** [comment]
When I said diff I meant the text diff.  I don't think the hatch changing width is real, that looks like issues with the renderer.  Do they move around if you pan / zoom / resize the window?

**Comment by afvincent** [comment]
The answer seems to be no with 1.5.1 (left panels) and yes with 2.0 (right panels):
![eps_and_hatches_differences_between_1 5 1_and_2 0_zoom_impact](https://cloud.githubusercontent.com/assets/17270724/24015041/6fd4d99e-0a87-11e7-8050-c859056e00e3.png)
The three level of zoom are 0.707, 1.000 and 2.000. With the EPS generated by Matplotlib 2.0, one can observe “glitches” for the two levels of zoom that are different from 1.

I agree that is is likely to be an issue with how the GS/Evince/Preview do the rendering.

I put the EPS files and a diff [there](https://gist.github.com/afvincent/735c866daa81e1f7eff6ec8299481f75)

**Comment by master-bit** [comment]
Hi, I've got a similar problem. I generate a hatched part in a figure with fill_between. Figure is saved to eps. It shows in Linux Document viewer and Okular. 
If I now incorparate the file into a pdf document via latex, the hatching is only visible in ocular, not in the document viewer. Also not printed.

![bildschirmfoto vom 2017-11-22 12 04 26](https://user-images.githubusercontent.com/19427706/33124475-6a0c7222-cf7d-11e7-9d18-847322a4d8e7.png)
first picture: The file opened in document viewer individually as eps (left) and in the pdf (right). Hatching not in the pdf

![bildschirmfoto vom 2017-11-22 12 05 06](https://user-images.githubusercontent.com/19427706/33124476-6a2a2092-cf7d-11e7-8708-27a5ad5bbe0b.png)
second picture: The file opened in okular individually as eps (left) and in the pdf (right). Hatching visible

**Comment by tacaswell** [comment]
Punting to 3.0, this is not going to get fixed in the next few days.

**Comment by tacaswell** [comment]
I suspect that the resolution here is going to involve someone fully understanding the eps/pdf spec and then sorting out which of the viewers is buggy....

This reminds me of the clipping issues we hand with librsvg for a while (we were outputting svg that complied with the spec, but rsvg did not render them correctly).

**Comment by mryodo** [comment]
Guys, has this issue been resolved? It seems like I'm having this problem on the latest module version. Or maybe any workaround?

**Comment by mksegv** [comment]
@mryodo, I got the behaviour correct by inserting `gsave` and `grestore` in `backend_ps.py`. I don't known whether this will incur other problems, but at least you could give it a try.

I don't know how to present my modification here, so I just attach my `git diff` below. Hope it helps.
```git
diff --git a/lib/matplotlib/backends/backend_ps.py b/lib/matplotlib/backends/backend_ps.py
index 62edbb919..c8877e2ab 100644
--- a/lib/matplotlib/backends/backend_ps.py
+++ b/lib/matplotlib/backends/backend_ps.py
@@ -311,7 +311,9 @@ class RendererPS(RendererBase):
             self._convert_path(Path.hatch(hatch), Affine2D().scale(sidelen),
                                simplify=False))
         self._pswriter.write("""\
+        gsave
         fill
+        grestore
         stroke
      } bind
    >>
```

**Comment by anntzer** [comment]
https://stackoverflow.com/questions/6614207/how-to-export-non-blurry-eps-images for another case of Preview apparently not rendering postscript correctly.

**Comment by mryodo** [comment]
Thanx! Works perfectly!

> @mryodo, I got the behaviour correct by inserting `gsave` and `grestore` in `backend_ps.py`. I don't known whether this will incur other problems, but at least you could give it a try.
> 
> I don't know how to present my modification here, so I just attach my `git diff` below. Hope it helps.
> 
> ```
> diff --git a/lib/matplotlib/backends/backend_ps.py b/lib/matplotlib/backends/backend_ps.py
> index 62edbb919..c8877e2ab 100644
> --- a/lib/matplotlib/backends/backend_ps.py
> +++ b/lib/matplotlib/backends/backend_ps.py
> @@ -311,7 +311,9 @@ class RendererPS(RendererBase):
>              self._convert_path(Path.hatch(hatch), Affine2D().scale(sidelen),
>                                 simplify=False))
>          self._pswriter.write("""\
> +        gsave
>          fill
> +        grestore
>          stroke
>       } bind
>     >>
> ```

**Comment by czeni** [comment]
Issue is still present in Matplotlib 3.1.1, but the posted fix worked like a charm! Thanks.

> @mryodo, I got the behaviour correct by inserting `gsave` and `grestore` in `backend_ps.py`. I don't known whether this will incur other problems, but at least you could give it a try.
> 
> I don't know how to present my modification here, so I just attach my `git diff` below. Hope it helps.
> 
> ```
> diff --git a/lib/matplotlib/backends/backend_ps.py b/lib/matplotlib/backends/backend_ps.py
> index 62edbb919..c8877e2ab 100644
> --- a/lib/matplotlib/backends/backend_ps.py
> +++ b/lib/matplotlib/backends/backend_ps.py
> @@ -311,7 +311,9 @@ class RendererPS(RendererBase):
>              self._convert_path(Path.hatch(hatch), Affine2D().scale(sidelen),
>                                 simplify=False))
>          self._pswriter.write("""\
> +        gsave
>          fill
> +        grestore
>          stroke
>       } bind
>     >>
> ```

**Comment by jklymak** [comment]
Does anyone want to open a PR for that change?

**Comment by QuLogic** [comment]
This should have been fixed by #15064.
